plugin/src/main/java/io/jenkins/plugins/coverage/metrics/steps/CoverageChecksPublisher.java

```diff
@@ -0,0 +1,419 @@
+                .withOutput(output)
+                .build();
+    }
+
+    private String getChecksTitle() {
+        return String.format("%s: %s",
+                FORMATTER.getDisplayName(Baseline.MODIFIED_LINES),
+                action.formatValue(Baseline.MODIFIED_LINES, Metric.LINE));
+    }
+
+    private String getSummary() {
+        var root = action.getResult();
+        return getOverallCoverageSummary(root) + "\n\n"
+                + getHealthReportSummary() + "\n\n"
+                + getProjectMetricsSummary(root);
+    }
+
+    private List<ChecksAnnotation> getAnnotations() {
+        var annotations = new ArrayList<ChecksAnnotation>();
+        for (var fileNode : action.getResult().filterChanges().getAllFileNodes()) {
+            for (var aggregatedLines : getAggregatedMissingLines(fileNode)) {
+                ChecksAnnotationBuilder builder = new ChecksAnnotationBuilder()
+                        .withPath(fileNode.getPath())
+                        .withTitle(Messages.Checks_Annotation_Title())
+                        .withAnnotationLevel(ChecksAnnotationLevel.WARNING)
+                        .withMessage(getAnnotationMessage(aggregatedLines))
+                        .withStartLine(aggregatedLines.startLine)
+                        .withEndLine(aggregatedLines.endLine);
+
+                annotations.add(builder.build());
+            }
+        }
+
+        return annotations;
+    }
+
+    private String getAnnotationMessage(final AggregatedMissingLines aggregatedMissingLines) {
+        if (aggregatedMissingLines.startLine == aggregatedMissingLines.endLine) {
+            return Messages.Checks_Annotation_Message_SingleLine(aggregatedMissingLines.startLine);
+        }
+        return Messages.Checks_Annotation_Message_MultiLine(aggregatedMissingLines.startLine,
+                aggregatedMissingLines.endLine);
+    }
+
+    private List<AggregatedMissingLines> getAggregatedMissingLines(final FileNode fileNode) {
+        var aggregatedMissingLines = new ArrayList<AggregatedMissingLines>();
+
+        if (fileNode.hasCoveredLinesInChangeSet()) {
+            var linesWithCoverage = fileNode.getLinesWithCoverage();
+            // there has to be at least one line when it is a file node with changes
+            var previousLine = linesWithCoverage.first();
+            var aggregatedLines = new AggregatedMissingLines(previousLine);
+            linesWithCoverage.remove(previousLine);
+
+            for (final var line : linesWithCoverage) {
+                if (line == previousLine + 1) {
+                    aggregatedLines.increaseEndLine();
+                }
+                else {
+                    aggregatedMissingLines.add(aggregatedLines);
+                    aggregatedLines = new AggregatedMissingLines(line);
+                }
+                previousLine = line;
+            }
+
+            aggregatedMissingLines.add(aggregatedLines);
+        }
+
+        return aggregatedMissingLines;
+    }
```

Comment:
Wouldn't it be more object oriented if that code is part of the model? So we should use something like: 

`fileNode.getMissingLines()`, `fileNode.getMissingBranches()`, or `fileNode.getMissingMutations()`?

Reply:
Hm the intention of this method is to provide groups of consecutive lines without coverage for the check annotations.
I am not sure if this would be to specific since it is only used here in that special case... In my opinion, it can stay here, but moving it would be okay for me too.

Reply:
Ok, then lets keep it as it is for now. I am still re-thinking on how to improve the model to make the decision of coverage vs. mutation coverage more object oriented. Currently we have a lot of code that checks if mutation coverage or line or branch coverage (source code painting, etc.).